src/jmespath.net/Expressions/JmesPathExpression.cs

```diff
@@ -8,8 +8,11 @@ namespace DevLab.JmesPath.Expressions
     /// <summary>
     /// Represents the base class for a JmesPath expression.
     /// </summary>
-    public abstract class JmesPathExpression
+    public abstract class JmesPathExpression : IContextHolder
     {
+        internal IScopeParticipant scopes_;
+        protected JToken context_;

```

Comment:
Maybe the `context_` should be scoped to the new `JmesPathExpressionTypeExpression` object, instead of _all_ expressions. I think it only makes sense for expression types.